### PR TITLE
Additional managed node outputs

### PR DESCRIPTION
**Pull request by TBeijen**
# PR o'clock

## Description

Work in progress for #643 

Additional node group output.

### Checklist

- [ ] Change added to CHANGELOG.md. All changes must be added and breaking changes and highlighted
- [ ] CI tests are passing
- [ ] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by TBeijen** [comment]
Sample output in current form:
```
eks_cluster_apps1_node_groups = [
  {
    "arn" = "arn:aws:eks:eu-west-1:NNNNNNNN:nodegroup/apps-1/apps-1-ng-managed-a-rapid-skylark/nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
    "iam_role_arn" = "arn:aws:iam::NNNNNNNN:role/apps-1-managed-node-groups"
    "id" = "arn:aws:eks:eu-west-1:NNNNNNNN:nodegroup/apps-1/apps-1-ng-managed-a-rapid-skylark/nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
    "name" = "apps-1-ng-managed-a-rapid-skylark"
    "resources" = [
      {
        "autoscaling_groups" = [
          {
            "name" = "eks-nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
          },
        ]
        "remote_access_security_group_id" = "sg-nnnnnnnnnnn"
      },
    ]
    "status" = "ACTIVE"
  },
  {
    "arn" = "arn:aws:eks:eu-west-1:NNNNNNNN:nodegroup/apps-1/apps-1-ng-managed-b-wired-parrot/nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
    "iam_role_arn" = "arn:aws:iam::NNNNNNNN:role/apps-1-managed-node-groups"
    "id" = "arn:aws:eks:eu-west-1:NNNNNNNN:nodegroup/apps-1/apps-1-ng-managed-b-wired-parrot/nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
    "name" = "apps-1-ng-managed-b-wired-parrot"
    "resources" = [
      {
        "autoscaling_groups" = [
          {
            "name" = "eks-nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
          },
        ]
        "remote_access_security_group_id" = "sg-nnnnnnnnnnn"
      },
    ]
    "status" = "ACTIVE"
  },
]
```

Same order as node group definition.

Would this format be favorable over having a list for each attribute, like:
```
node_group_arns = [
    "arn:aws:eks:eu-west-1:NNNNNNNN:nodegroup/apps-1/apps-1-ng-managed-a-rapid-skylark/nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn",
    "arn:aws:eks:eu-west-1:NNNNNNNN:nodegroup/apps-1/apps-1-ng-managed-b-wired-parrot/nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn"
}
node_group_iam_role_arns = [
     "arn:aws:iam::NNNNNNNN:role/apps-1-managed-node-groups",
    "arn:aws:iam::NNNNNNNN:role/apps-1-managed-node-groups"
]
# etc...
```

The latter might be more consistent with existing worker group outputs but I wonder what would be most user-friendly. Thoughts?

File: outputs.tf [hunk 1]
```diff
@@ -165,3 +165,19 @@ output "node_groups_iam_role_arns" {
     node_group.node_group_name => node_group.node_role_arn
   }
 }
+
+output "node_groups" {
+  value = [
```
Comment (barryib) [inline outputs.tf]: could you please add description ?

Comment (TBeijen) [reply outputs.tf]: Sure. Was planning on adding a docs `.md` describing matters in more detail. First I wanted to see if there's feedback on the output structure.

**Comment by dpiddockcmp** [comment]
The node_groups themselves are generated from a `for_each`. I feel this output would be much easier to work with if it was a map of maps instead of a list of maps. Similar to the `node_groups_iam_role_arns` output just above. The node_groups don't have a predictable order due to being created from a `for_each`.

And now I've looked at the code a little closer, the `node_group.node_group_name` is generated based on a random pet resource which makes that output pretty much useless for predictable reuse. 🤷‍♂

This node_group code is confusing. Why is user input a list of maps, which then get turned into a map of maps based on something that's not enforced unique in the user input? Oh my.

Totally out of scope of this PR but I'd love for this to be input:
```hcl
node_groups = {
  group1 = { instance_type = "t3.medium" }
  group2 = { instance_type = "c5.xlarge" }
}
```
And your outputs to be:
```hcl
node_groups = {
  group1 = {
    role_arn = "..."
    ...
  }
  group2 = {
    role_arn = "..."
    ...
  }
}
```

**Comment by TBeijen** [comment]
> The node_groups themselves are generated from a for_each. I feel this output would be much easier to work with if it was a map of maps instead of a list of maps.

> This node_group code is confusing. Why is user input a list of maps, which then get turned into a map of maps based on something that's not enforced unique in the user input? Oh my.

I chose list output based on `node_groups` input being a list. But, admittedly I didn't take a very close look at the inner workings of the existing code, assuming (never assume) constructs were there for a reason.

Imo the intermediate map based on something not guaranteed unique is a red flag. 

In the light of @max-rocket-internet [comment](https://github.com/terraform-aws-modules/terraform-aws-eks/issues/635#issuecomment-568474677) in #635 I'm inclined to focus on a poc for a separate managed_node_group module instead. It might be less work than hammering this into a good shape and quite a lot of the complexity in the current state seems to originate from needing to manage lists/maps of inputs and outputs.

**Comment by dpiddockcmp** [comment]
Ran with my comments and created #650

**Comment by max-rocket-internet** [comment]
> I'm inclined to focus on a poc for a separate managed_node_group module instead

I think that's a good idea. I'll close this then 🙂

**Pull request closed without merging**